### PR TITLE
Refactor(eos_cli_config_gen): Improve schema for redistributes_routes under address_family_ipv4_multicast, address_family_ipv6 and vrfs for router_bgp

### PR DESCRIPTION
**Pull request by MaheshGSLAB**
## Change Summary

Improve schema for redistributes_routes under address_family_ipv4_multicast, address_family_ipv6 and vrfs for router_bgp

## Related Issue(s)

Fixes #<ISSUE ID>

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
Improve schema for redistributes_routes under address_family_ipv4_multicast, address_family_ipv6 and vrfs for router_bgp

## How to test
Run molecule

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

**Review by laxmikantchintakindi — DISMISSED** [review]

File: ansible_collections/arista/avd/docs/porting-guides/5.x.x.md [hunk 1]
```diff
@@ -285,6 +285,16 @@ As of AVD 5.0.0 "null" option for esp integrity and encryption has been replaced
 +      encryption: disabled
 ```
 
+### Added valid for data models `router_bgp.address_family_ipv4_multicast.redistribute_routes[].source_protocol`, `router_bgp.address_family_ipv6.redistribute_routes[].source_protocol`, and `router_bgp.vrf.redistribute_routes[].source_protocol`
+
+Added valid for data models `router_bgp.address_family_ipv4_multicast.redistribute_routes[].source_protocol`, `router_bgp.address_family_ipv6.redistribute_routes[].source_protocol`, and `router_bgp.vrf.redistribute_routes[].source_protocol` to restrict the source protocol of redistribute routes in EOS configuration.
```
Comment (Shivani-gslab) [inline ansible_collections/arista/avd/docs/porting-guides/5.x.x.md]:
```suggestion
### Added valid values for data models `router_bgp.address_family_ipv4_multicast.redistribute_routes[].source_protocol`, `router_bgp.address_family_ipv6.redistribute_routes[].source_protocol`, and `router_bgp.vrf.redistribute_routes[].source_protocol`

Added valid values for data models `router_bgp.address_family_ipv4_multicast.redistribute_routes[].source_protocol`, `router_bgp.address_family_ipv6.redistribute_routes[].source_protocol`, and `router_bgp.vrf.redistribute_routes[].source_protocol` to restrict the source protocol of redistribute routes in EOS configuration.
```

**Review by Shivani-gslab — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/docs/porting-guides/5.x.x.md (hunk 1 above)
Comment (laxmikantchintakindi) [reply ansible_collections/arista/avd/docs/porting-guides/5.x.x.md]: We can brief line 288 like - Added valid values check for `redistribute_routes[].source_protocol` in `router_bgp` data model.'

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 2]
```diff
@@ -1646,17 +1645,16 @@ keys:
               keys:
                 source_protocol:
                   type: str
-                  # TODO AVD5.0: Add valid values and ensure we cover all options.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml:2039]: "...and ensure we cover all options."

Did you do that?

Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml:2039]: yes all options are covered apart from isis levels which is pending for all type of redistribute routes
```
wan1-dc1(config-router-bgp-af)#redistribute isis ?
  include    Include following routes while redistributing
  level-1    Redistribute IS-IS level-1 routes <-----------
  level-1-2  Redistribute IS-IS level-1 and level-2 routes <-------------
  level-2    Redistribute IS-IS level-2 routes <------------
  rcf        Routing control functions configuration to filter and modify paths
  route-map  Name a route map
```

Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml:2039]: so if we add the valid_values now, someone using "isis level-2" would not have an alternative. We need to address that before we tighten the schema. I think we have this covered in some areas, so hopefully we can reuse some logic/models.

**Comment by MaheshGSLAB** [comment]
keep on hold for 5.x

**Comment by ClausHolbechArista** [comment]
@MaheshGSLAB please revive this PR :)

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md [hunk 3]
```diff
@@ -1782,6 +1824,73 @@
 
             # Outbound route-map name.
             route_map_out: <str>
+
+        # Redistribute routes in to BGP.
+        redistribute:
+          attached_host:
+            enabled: <bool>
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md:1831]: need to set enabled as required key

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md [hunk 4]
```diff
@@ -1782,6 +1824,73 @@
...
+
+            # OSPF routes learned from internal sources.
+            match_internal:
+              enabled: <bool>
+              route_map: <str>
+
+            # OSPF routes learned from external NSSA sources.
+            match_nssa_external:
+              enabled: <bool>
+              nssa_type: <int>
```
Comment (MaheshGSLAB) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/tables/router-bgp.md:1885]: need to add the support of nssa_type in j2 template

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 5]
```diff
@@ -1862,23 +1899,157 @@ keys:
...
+                      `route_map` and `rcf` are mutually exclusive. `route_map` takes precedence.
+                  include_leaked:
+                    description: Include following routes while redistributing.
+                    type: bool
+              ospf:
+                type: dict
+                keys:
+                  enabled:
+                    type: bool
+                    description: Generate redistribute ospf with or without mathc internal, include leaked and route map.
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]:
```suggestion
                    description: Generate redistribute ospf with or without match internal, include leaked and route map.
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 6]
```diff
@@ -1862,23 +1899,157 @@ keys:
...
+                      route_map:
+                        type: str
+                  route_map:
+                    type: str
+              ospfv3:
+                type: dict
+                keys:
+                  enabled:
+                    type: bool
+                    description: Generate redistribute ospf with or without mathc internal, include leaked and route map.
```
Comment (Shivani-gslab) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]:
```suggestion
                    description: Generate redistribute ospf with or without match internal, include leaked and route map.
```

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2 [hunk 7]
```diff
@@ -1268,25 +1278,128 @@ router bgp {{ router_bgp.as }}
...
+{%                 elif redistribute_var.ospf.match_internal.enabled is arista.avd.defined(true) %}
+{%                     set redistribute_ospf = redistribute_ospf ~ " match internal" %}
+{%                     if redistribute_var.ospf.match_internal.route_map is arista.avd.defined %}
+{%                         set redistribute_ospf = redistribute_ospf ~ " route-map " ~ redistribute_var.ospf.match_internal.route_map %}
 {%                     endif %}
 {%                 endif %}
-{%                 if redistribute_route.include_leaked is arista.avd.defined and redistribute_route.source_protocol == "isis" %}
-{%                     set redistribute_route_cli = redistribute_route_cli ~ " include leaked" %}
+      {{ redistribute_ospf }}
+{%             endif %}
```
Comment (Shivani-gslab) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2]: here `redistribute_var.ospf.enabled` is not a `required key` and as per the logic here `match_internal` could not be configured if `redistribute_var.ospf.enabled is not true`
same at other places

Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2]: updated with or condition

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2 [hunk 8]
```diff
@@ -1385,29 +1498,162 @@ router bgp {{ router_bgp.as }}
...
+{%                     endif %}
+{%                 elif redistribute_var.ospfv3.match_internal.enabled is arista.avd.defined(true) %}
+{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " match internal" %}
+{%                     if redistribute_var.ospfv3.match_internal.include_leaked is arista.avd.defined(true) %}
+{%                         set redistribute_ospfv3 = redistribute_ospfv3 ~ " include leaked" %}
+{%                     endif %}
+{%                     if redistribute_var.ospfv3.match_internal.route_map is arista.avd.defined %}
+{%                         set redistribute_ospfv3 = redistribute_ospfv3 ~ " route-map " ~ redistribute_var.ospfv3.match_internal.route_map %}
 {%                     endif %}
 {%                 endif %}
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2:1575]:
```suggestion
{%             if redistribute_var.ospfv3.match_internal.enabled is arista.avd.defined(true) %}
{%                 set redistribute_ospfv3 = "redistribute ospfv3 match internal" %}
{%                 if redistribute_var.ospfv3.match_internal.include_leaked is arista.avd.defined(true) %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " include leaked" %}
{%                 endif %}
{%                 if redistribute_var.ospfv3.match_internal.route_map is arista.avd.defined %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " route-map " ~ redistribute_var.ospfv3.match_internal.route_map %}
{%                 endif %}
      {{ redistribute_ospfv3 }}
{%             elif redistribute_var.ospfv3.enabled is arista.avd.defined(true) %}
{%                 set redistribute_ospfv3 = "redistribute ospfv3" %}
{%                 if redistribute_var.ospfv3.include_leaked is arista.avd.defined(true) %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " include leaked" %}
{%                 endif %}
{%                 if redistribute_var.ospfv3.route_map is arista.avd.defined %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " route-map " ~ redistribute_var.ospfv3.route_map %}
{%                 endif %}
      {{ redistribute_ospfv3 }}
{%             endif %}
```

Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2]:
```suggestion
{%             if redistribute_var.ospfv3.match_internal.enabled is arista.avd.defined(true) %}
{%                 set redistribute_ospfv3 = "redistribute ospfv3 match internal" %}
{%                 if redistribute_var.ospfv3.match_internal.include_leaked is arista.avd.defined(true) %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " include leaked" %}
{%                 endif %}
{%                 if redistribute_var.ospfv3.match_internal.route_map is arista.avd.defined %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " route-map " ~ redistribute_var.ospfv3.match_internal.route_map %}
{%                 endif %}
      {{ redistribute_ospfv3 }}
{%             elif redistribute_var.ospfv3.enabled is arista.avd.defined(true) %}
{%                 set redistribute_ospfv3 = "redistribute ospfv3" %}
{%                 if redistribute_var.ospfv3.include_leaked is arista.avd.defined(true) %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " include leaked" %}
{%                 endif %}
{%                 if redistribute_var.ospfv3.route_map is arista.avd.defined %}
{%                     set redistribute_ospfv3 = redistribute_ospfv3 ~ " route-map " ~ redistribute_var.ospfv3.route_map %}
{%                 endif %}
```

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2 [hunk 9]
```diff
@@ -2025,29 +2276,196 @@ router bgp {{ router_bgp.as }}
...
+{%                 if redistribute_var.ospf.match_nssa_external.include_leaked is arista.avd.defined(true) %}
+{%                     set redistribute_ospf = redistribute_ospf ~ " include leaked" %}
+{%                 endif %}
+{%                 if redistribute_var.ospf.match_nssa_external.route_map is arista.avd.defined %}
+{%                     set redistribute_ospf_match = redistribute_ospf_match ~ " route-map " ~ redistribute_var.ospf.match_nssa_external.route_map %}
+{%                 endif %}
+      {{ redistribute_ospf_match }}
+{%             endif %}
+{%             if redistribute_var.ospfv3.enabled is arista.avd.defined(true) or redistribute_var.ospfv3.match_internal.enabled is arista.avd.defined(true) %}
+{%                 set redistribute_ospfv3 = "redistribute ospfv3" %}
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2:2369]: Need to add similar logic as comment on address_family_ipv6 redistribute.ospfv3

Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2:2369]: why not just make the enabled key required?

Comment (ClausHolbechArista) [reply python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2:2369]: I think I understand it now. The outer `enabled` is for the general redistribute ...

File: python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2 [hunk 10]
```diff
@@ -2025,29 +2276,196 @@ router bgp {{ router_bgp.as }}
...
+{%                     set redistribute_isis = redistribute_isis ~ " include leaked" %}
+{%                 endif %}
+{%                 if redistribute_var.isis.route_map is arista.avd.defined %}
+{%                     set redistribute_isis = redistribute_isis ~ " route-map " ~ redistribute_var.isis.route_map %}
+{%                 elif redistribute_var.isis.rcf is arista.avd.defined %}
+{%                     set redistribute_isis = redistribute_isis ~ " rcf " ~ redistribute_var.isis.rcf %}
+{%                 endif %}
+      {{ redistribute_isis }}
+{%             endif %}
+{%             if redistribute_var.ospf.enabled is arista.avd.defined(true) or redistribute_var.ospf.match_internal.enabled is arista.avd.defined(true) %}
```
Comment (laxmikantchintakindi) [inline python-avd/pyavd/_eos_cli_config_gen/j2templates/eos/router-bgp.j2]: Needs add similar lines as commented on address_family_ipv6. redistribute.ospv3.

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 11]
```diff
@@ -1862,23 +1899,157 @@ keys:
...
+                      route_map:
+                        type: str
+                  route_map:
+                    type: str
+              ospfv3:
+                type: dict
+                keys:
+                  enabled:
+                    type: bool
+                    description: Generate redistribute ospf with or without match internal, include leaked and route map.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]:
```suggestion
                    description: Redistribute OSPFv3 routes.
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 12]
```diff
@@ -1862,23 +1899,157 @@ keys:
...
+                  match_external:
+                    description: OSPF routes learned from external sources.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]:
```suggestion
                    description: Redistribute OSPFv3 routes learned from external sources.
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 13]
```diff
@@ -1862,23 +1899,157 @@ keys:
...
+                    type: dict
+                    keys:
+                      enabled:
+                        type: bool
+                        required: true
+                      route_map:
+                        type: str
+                  match_internal:
+                    description: OSPF routes learned from internal sources.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]:
```suggestion
                    description: Redistribute OSPFv3 routes learned from internal sources.
```

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml [hunk 14]
```diff
@@ -1862,23 +1899,157 @@ keys:
...
+                    type: dict
+                    keys:
+                      enabled:
+                        type: bool
+                        required: true
+                      route_map:
+                        type: str
+                  match_nssa_external:
+                    description: OSPF routes learned from external NSSA sources.
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]:
```suggestion
                    description: Redistribute OSPFv3 routes learned from external NSSA sources.
```

**Review by ClausHolbechArista — COMMENTED** [review]
Same comments for other contexts.

**Review by Shivani-gslab — APPROVED** [review]
LGTM

File: python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml (hunk 11 above)
Comment (MaheshGSLAB) [reply python-avd/pyavd/_eos_cli_config_gen/schema/schema_fragments/router_bgp.schema.yml]: updated all over

**Review by ClausHolbechArista — APPROVED** [review]

**Pull request merged**